compiler/transforms/accfg_config_overlap.py

```diff
@@ -0,0 +1,74 @@
+from xdsl.dialects import builtin
+from xdsl.ir import MLContext, Operation, Use
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+from compiler.dialects import accfg
+
+
+class BlockLevelSetupAwaitOverlapPattern(RewritePattern):
+    """
+    Inspects a single block for the following structure:
+
+    ```
+     %token = accfg.launch %state ...
+     // ...
+     %next_state = accfg.setup ... from %state ...
+    ```
+
+    And moves the setup op right behind the launch, iff:
+    - The these two ops are the only two uses of the `%state` SSA value
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: accfg.SetupOp, rewriter: PatternRewriter, /):
+        # ignore setup without an input state
+        if not op.in_state:
+            return
+        # only apply if there are exactly two uses of the state
+        ops_uses = get_ops_from_uses(op.in_state.uses)
+        if len(ops_uses) != 2:
+            return
+        # check that one use is this op, and the other is a launch op
+        # and then grab a reference to the launch op
+        op1, op2 = ops_uses
+        launch: accfg.LaunchOp | None = None
+        if op1 == op and isinstance(op2, accfg.LaunchOp):
+            launch = op2
+        elif op2 == op and isinstance(op1, accfg.LaunchOp):
+            launch = op1
+        else:
+            return
+
+        # check that the launch and the setup are on the same block level
+        if launch.parent_block() != op.parent_block():
+            return
+
+        # abort if the launch op has already been moved (prevent forever loops)
+        if launch.next_op is op:
+            return
```

Comment:
Maybe it makes more sense to put this on the top as the first check?

Reply:
We can only do this after we have the reference to the `launch`, so can only move it up one check, but I don't think that matters much.

Reply:
Right! Ok!